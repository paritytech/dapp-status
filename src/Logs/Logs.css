/* Copyright 2015-2017 Parity Technologies (UK) Ltd.
/* This file is part of Parity.
/*
/* Parity is free software: you can redistribute it and/or modify
/* it under the terms of the GNU General Public License as published by
/* the Free Software Foundation, either version 3 of the License, or
/* (at your option) any later version.
/*
/* Parity is distributed in the hope that it will be useful,
/* but WITHOUT ANY WARRANTY; without even the implied warranty of
/* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
/* GNU General Public License for more details.
/*
/* You should have received a copy of the GNU General Public License
/* along with Parity.  If not, see <http://www.gnu.org/licenses/>.
 */

:local(.settingsButton) {
  box-shadow: none !important;
  padding: 0.2em !important;
}

:local(.settingsButton):first-of-type {
  margin-left: 1em !important;
}

:local(.log) {
  font-family: monospace;
  font-size: 0.9em;
  white-space: pre-line;
  word-wrap: break-word;
  color: #333;
  margin: 0.5em 0;
  display: flex;
  align-items: center;
}

:local(.logDate) {
  color: green;
  font-size: 0.85em;
  margin-right: 0.5em;
}

:local(.logText) {
  flex: 1;
}
