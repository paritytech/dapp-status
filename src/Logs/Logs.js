// Copyright 2015-2017 Parity Technologies (UK) Ltd.
// This file is part of Parity.

// Parity is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.

// Parity is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.

// You should have received a copy of the GNU General Public License
// along with Parity.  If not, see <http://www.gnu.org/licenses/>.

import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { inject, observer } from 'mobx-react';
import Button from 'semantic-ui-react/dist/commonjs/elements/Button';
import Header from 'semantic-ui-react/dist/commonjs/elements/Header';
import { FormattedMessage } from 'react-intl';

import Section from '../components/Section';
import styles from './Logs.css';

class Logs extends Component {
  state = {
    isLogging: false,
    sort: 'descending'
  };

  static propTypes = {
    devLogsLevelsStore: PropTypes.object.isRequired,
    devLogsStore: PropTypes.object.isRequired
  };

  handleToggleLogging = () =>
    this.setState({ isLogging: !this.state.isLogging });

  handleToggleSort = () =>
    this.setState({
      sort: this.state.sort === 'ascending' ? 'descending' : 'ascending'
    });

  renderLogs = () => {
    const { sort } = this.state;
    const { devLogsStore } = this.props;

    const logs =
      sort === 'descending'
        ? devLogsStore.parsedLogs.slice(0, 25).reverse()
        : devLogsStore.parsedLogs.slice(0, 25);

    return logs.map((log, index) => (
      <p key={index} className={styles.log}>
        <span className={styles.logDate}>[{log.date.toLocaleString()}]</span>
        <span className={styles.logText}>{log.log}</span>
      </p>
    ));
  };

  render() {
    const { devLogsLevelsStore } = this.props;

    return (
      <Section
        title={
          <div>
            <FormattedMessage
              as={Header.Content}
              id="dapp.status.logs.title"
              defaultMessage="Node Logs"
            />
            <Header.Subheader>
              <FormattedMessage
                as={Header.Content}
                id="dapp.status.logs.subtitle"
                defaultMessage="Debug Level: {level}"
                values={{ level: devLogsLevelsStore.devLogsLevel || 'Default' }}
              />
              <Button
                basic
                circular
                className={styles.settingsButton}
                compact
                icon={this.state.isLogging ? 'pause' : 'play'}
                onClick={this.handleToggleLogging}
                size="large"
              />
              <Button
                basic
                circular
                className={styles.settingsButton}
                compact
                icon={`sort numeric ${this.state.sort}`}
                onClick={this.handleToggleSort}
                size="large"
              />
            </Header.Subheader>
          </div>
        }
      >
        {this.state.isLogging ? (
          this.renderLogs()
        ) : (
          <FormattedMessage
            id="dapp.status.logs.isLogging"
            defaultMessage="Refresh and display of logs from Parity is currently stopped via the UI, start it to see the latest updates."
          />
        )}
      </Section>
    );
  }
}

export default inject('devLogsLevelsStore', 'devLogsStore')(observer(Logs));
