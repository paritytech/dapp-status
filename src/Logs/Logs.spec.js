// Copyright 2015-2017 Parity Technologies (UK) Ltd.
// This file is part of Parity.

// Parity is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.

// Parity is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.

// You should have received a copy of the GNU General Public License
// along with Parity.  If not, see <http://www.gnu.org/licenses/>.

import React from 'react';
import { shallow } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';
import Button from 'semantic-ui-react/dist/commonjs/elements/Button';

import { mountWithIntl } from '../setupTests';
import { Logs } from './Logs';

const props = {
  devLogsStore: {
    parsedLogs: [
      { date: '1-1-1', log: 'FooLog' },
      { date: '2-2-2', log: 'BarLog' }
    ]
  },
  devLogsLevelsStore: {
    devLogsLevels: 'Baz'
  }
};

test('should render correctly when not logging', () => {
  const component = shallow(<Logs {...props} />);

  expect(shallowToJson(component)).toMatchSnapshot();
});

test('should render correctly when logging', () => {
  const component = shallow(<Logs {...props} />);
  component.setState({ isLogging: true });

  expect(shallowToJson(component)).toMatchSnapshot();
});

test('should render correctly when sort ascending', () => {
  const component = shallow(<Logs {...props} />);
  component.setState({ isLogging: true, sort: 'ascending' });

  expect(shallowToJson(component)).toMatchSnapshot();
});

test('should handle toggle logging', () => {
  const component = mountWithIntl(<Logs {...props} />);
  const button = component.find(Button).first();

  button.props().onClick();
  expect(component.state('isLogging')).toBe(true);

  button.props().onClick();
  expect(component.state('isLogging')).toBe(false);
});

test('should handle toggle sort', () => {
  const component = mountWithIntl(<Logs {...props} />);
  const button = component.find(Button).at(1);

  button.props().onClick();
  expect(component.state('sort')).toBe('ascending');

  button.props().onClick();
  expect(component.state('sort')).toBe('descending');
});
