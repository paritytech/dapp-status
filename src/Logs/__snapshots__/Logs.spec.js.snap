// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly when logging 1`] = `
<Section
  title={
    <div>
      <FormattedMessage
        as={[Function]}
        defaultMessage="Node Logs"
        id="dapp.status.logs.title"
        values={Object {}}
      />
      <HeaderSubheader>
        <FormattedMessage
          as={[Function]}
          defaultMessage="Debug Level: {level}"
          id="dapp.status.logs.subtitle"
          values={
            Object {
              "level": "Default",
            }
          }
        />
        <Button
          as="button"
          basic={true}
          circular={true}
          className={undefined}
          compact={true}
          icon="pause"
          onClick={[Function]}
          size="large"
        />
        <Button
          as="button"
          basic={true}
          circular={true}
          className={undefined}
          compact={true}
          icon="sort numeric descending"
          onClick={[Function]}
          size="large"
        />
      </HeaderSubheader>
    </div>
  }
>
  <p
    key="0"
  >
    <span>
      [
      2-2-2
      ]
    </span>
    <span>
      BarLog
    </span>
  </p>
  <p
    key="1"
  >
    <span>
      [
      1-1-1
      ]
    </span>
    <span>
      FooLog
    </span>
  </p>
</Section>
`;

exports[`should render correctly when not logging 1`] = `
<Section
  title={
    <div>
      <FormattedMessage
        as={[Function]}
        defaultMessage="Node Logs"
        id="dapp.status.logs.title"
        values={Object {}}
      />
      <HeaderSubheader>
        <FormattedMessage
          as={[Function]}
          defaultMessage="Debug Level: {level}"
          id="dapp.status.logs.subtitle"
          values={
            Object {
              "level": "Default",
            }
          }
        />
        <Button
          as="button"
          basic={true}
          circular={true}
          className={undefined}
          compact={true}
          icon="play"
          onClick={[Function]}
          size="large"
        />
        <Button
          as="button"
          basic={true}
          circular={true}
          className={undefined}
          compact={true}
          icon="sort numeric descending"
          onClick={[Function]}
          size="large"
        />
      </HeaderSubheader>
    </div>
  }
>
  <FormattedMessage
    defaultMessage="Refresh and display of logs from Parity is currently stopped via the UI, start it to see the latest updates."
    id="dapp.status.logs.isLogging"
    values={Object {}}
  />
</Section>
`;

exports[`should render correctly when sort ascending 1`] = `
<Section
  title={
    <div>
      <FormattedMessage
        as={[Function]}
        defaultMessage="Node Logs"
        id="dapp.status.logs.title"
        values={Object {}}
      />
      <HeaderSubheader>
        <FormattedMessage
          as={[Function]}
          defaultMessage="Debug Level: {level}"
          id="dapp.status.logs.subtitle"
          values={
            Object {
              "level": "Default",
            }
          }
        />
        <Button
          as="button"
          basic={true}
          circular={true}
          className={undefined}
          compact={true}
          icon="pause"
          onClick={[Function]}
          size="large"
        />
        <Button
          as="button"
          basic={true}
          circular={true}
          className={undefined}
          compact={true}
          icon="sort numeric ascending"
          onClick={[Function]}
          size="large"
        />
      </HeaderSubheader>
    </div>
  }
>
  <p
    key="0"
  >
    <span>
      [
      1-1-1
      ]
    </span>
    <span>
      FooLog
    </span>
  </p>
  <p
    key="1"
  >
    <span>
      [
      2-2-2
      ]
    </span>
    <span>
      BarLog
    </span>
  </p>
</Section>
`;
